meadow-client/src/Halogen/Blockly.purs

```diff
@@ -0,0 +1,123 @@
+module Halogen.Blockly where
+
+import Blockly (BlockDefinition, Div(..))
+import Blockly as Blockly
+import Blockly.Generator (Generator, newBlock, workspaceToCode)
+import Blockly.Types as BT
+import Bootstrap (btn, btnInfo, btnSmall)
+import Control.Alt ((<|>))
+import Data.Either (Either(..))
+import Data.Maybe (Maybe(..))
+import Data.NaturalTransformation (type (~>))
+import Effect.Class (class MonadEffect)
+import Effect.Class.Console (log)
+import Halogen (ClassName(..), Component, ComponentDSL, ComponentHTML, RefLabel(..), action, get, lifecycleComponent, liftEffect, modify, raise)
+import Halogen.HTML (button, div, text)
+import Halogen.HTML as HH
+import Halogen.HTML.Events (input_, onClick)
+import Halogen.HTML.Events as HE
+import Halogen.HTML.Properties (classes, id_, ref)
+import Marlowe.Blockly (buildBlocks, buildGenerator)
+import Marlowe.Parser as Parser
+import Marlowe.Pretty (pretty)
+import Marlowe.Types (Contract(..))
+import Prelude (Unit, bind, discard, pure, show, unit, ($))
+import Text.Parsing.Parser (runParser)
+import Text.Parsing.Parser.Basic (parens)
+
+type BlocklyState
+  = {blocklyState :: Maybe BT.BlocklyState, generator :: Maybe Generator}
+
+data BlocklyQuery a
+  = Inject (Array BlockDefinition) a
+  | Resize a
+  | SetData Unit a
+  | GetCode a
+  | SetCode String a
+
+data BlocklyMessage
+  = Initialized
+  | CurrentCode String
+
+type HTML
+  = ComponentHTML BlocklyQuery
+
+type DSL m
+  = ComponentDSL BlocklyState BlocklyQuery BlocklyMessage m
+
+blockly :: forall m. MonadEffect m => Array BlockDefinition -> Component HH.HTML BlocklyQuery Unit BlocklyMessage m
+blockly blockDefinitions =
+  lifecycleComponent
+    { initialState: \_ -> {blocklyState: Nothing, generator: Nothing}
+    , render
+    , eval
+    , initializer: Just $ action (Inject blockDefinitions)
+    , finalizer: Nothing
+    , receiver: HE.input SetData
+    }
+
+eval :: forall m. MonadEffect m => BlocklyQuery ~> DSL m
+eval (Inject blockDefinitions next) = do
+  blocklyState <- liftEffect $ Blockly.createBlocklyInstance (Div "blocklyWorkspace") (Div "blocklyToolbox")
```

Comment:
You use these IDs in a couple of places. I think it's worth extracting them as constants.

Reply:
no, they're only used in this place.